### PR TITLE
`POST` request for qr-code mobile auth

### LINKED ISSUE #941: Feature Request: POST Request to create a qr-code-auth document
It is used while scanning the QR code from the mobile app on my-site QR code.

**Request**
POST

**Endpoint**
 /auth/qr-code-auth
 
 **Request Body**
```
 {
  "device_info": "t5k77PHnuDSrgEzvMJAj",
  "user_id": "NLFSj7Kz30oHgolfIZtJ",
  "device_id": "sdkdjASdjdsfns3kdj"
}

```
**Request Response**
Status: 201
Content-Type: application/json

```
{
  "data": {
    "user_id": "SooJK37gzjIZfFNH0tlL",
    "device_info": "t5k77PHnuDSrgEzvMJAj",
    "device_id": "sdkdjASdjdsfns3kdj",
    "authorization_status": "NOT_INIT"
  },
  "message": "Authentication document created successfully."
}
```
In case of any doubt let's have a discussion on a call or you can text @mobile-app-react-native channel.
You can also check our[ API contract](https://github.com/Real-Dev-Squad/website-api-contracts/tree/main/qr-code-auth
) in case of error responses.

### PR DESCRIPTION
**Pull request by isVivek99**
This `PR` closes #941  , and `1/3` for #1176
### Description:
this call initializes user in the DB, when the call is made.
Screenshot 2023-06-23 at 3 07 06 PM
---
Screenshot 2023-06-23 at 3 07 19 PM

### Test Coverage stats

File | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-- | -- | -- | -- | -- | --
All files | 69 |    57.92 |   64.83 |    69.2 
website-backend/controllers | 66.87 |  61.04 | 71.83 |  67 |        
qrCodeAuth.js | 84.61 |  100 | 100 |  84.61 | 19-20    
website-backend/middlewares/validators |   77.94 |    66.66 |   77.35 |   78.33 |  
 qrCodeAuth.js |   93.33 |      100 |     100 |   93.33 | 14    
 website-backend/models | 64.55 |  64.17 | 59.48 | 64.64 |   
  qrCodeAuth.js |   91.66 |       80 |     100 |   91.66 | 35-36 |
 website-backend/routes                 |     100 |      100 |     100 |     100 |   
  auth.js                               |     100 |      100 |     100 |     100 |

### CODE REVIEW

File: controllers/auth.js [hunk 1]
```diff
@@ -1,7 +1,11 @@
 const passport = require("passport");
 const users = require("../models/users");
+const QrCodeAuthModel = require("../models/qrCodeAuth");
 const authService = require("../services/authService");
-
+const { SOMETHING_WENT_WRONG } = require("../constants/errorMessages");
+const DATA_ADDED_SUCCESSFULLY = "User Device Info added successfully!";
```
Comment (harshith-venkatesh) [inline controllers/auth.js:6]: Would it be better to have separate file to save all constant messages
Need not be part of this PR
talking in general

File: controllers/auth.js [hunk 2]
```diff
@@ -66,7 +70,51 @@ const signout = (req, res) => {
...
+    }
+
+    return res.status(201).json({
+      ...userInfo,
+      message: DATA_ADDED_SUCCESSFULLY,
+    });
+  } catch (err) {
+    logger.error(`Error while storing user device info : ${err}`);
+    return res.boom.badImplementation(SOMETHING_WENT_WRONG);
+  }
```
Comment (heyrandhir) [inline controllers/auth.js:113]: Nit: Catch specific errors instead of a generic catch all. This helps with better error handling.

File: controllers/auth.js [hunk 3]
```diff
@@ -66,7 +70,51 @@ const signout = (req, res) => {
...
+      });
+    }
+
+    const userInfo = await QrCodeAuthModel.storeUserDeviceInfo(userJson);
+
+    if (!userInfo) {
+      return res.status(400).json({
+        message: BAD_REQUEST,
+      });
+    }
```
Comment (heyrandhir) [inline controllers/auth.js:104]: 400 bad request code is used for in general for malformed request. I think if `userInfo` is unfound we should throw 404.

File: controllers/users.js [hunk 4]
```diff
@@ -5,7 +5,6 @@ const logsQuery = require("../models/logs");
 const imageService = require("../services/imageService");
 const { profileDiffStatus } = require("../constants/profileDiff");
 const { logType } = require("../constants/logs");
-
```
Comment (heyrandhir) [inline controllers/users.js:8]: I think this shouldn't be part of this PR. You might have formatted accidentally.

File: middlewares/validators/userDeviceInfo.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+const joi = require("joi");
+
+const storeUserDeviceInfo = async (req, res, next) => {
+  const schema = joi.object().strict().keys({
+    user_id: joi.string().required(),
+    device_info: joi.string().required(),
+    device_id: joi.string().required(),
```
Comment (heyrandhir) [inline middlewares/validators/userDeviceInfo.js:7]: any specific reason for snake casing here. I have seen we are using camel Casing for Object key names at other places

File: models/qrCodeAuth.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * @param userDeviceInfoData { Object }: User device info data object to be stored in DB
+ * @return {Promise<{userDeviceInfoData|Object}>}
+ */
+const storeUserDeviceInfo = async (userDeviceInfoData) => {
+  try {
+    const { user_id: userId } = userDeviceInfoData;
+    const user = await userModel.doc(userId).get();
+    if (user.data()) {
+      await QrCodeAuthModel.doc(userId).set(userDeviceInfoData);
```
Comment (heyrandhir) [inline models/qrCodeAuth.js:16]: are we keeping the document id in this collection = userId ?

File: models/qrCodeAuth.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+  } catch (err) {
+    logger.error("Error in storing user device info.", err);
+    throw err;
+  }
+};
+
+const getUserAuthStatus = async (userId) => {
+  try {
+    const userAuthStatus = await QrCodeAuthModel.where("user_id", "==", userId).limit(1).get();
```
Comment (heyrandhir) [inline models/qrCodeAuth.js:30]: will we be having multiple documents belonging to a single user ?

File: test/integration/qrCodeAuth.test.js
Comment (heyrandhir) [inline test/integration/qrCodeAuth.test.js:1]: why integration test in 2 files 🤔

**Review by heyrandhir — CHANGES_REQUESTED** [review]
How will we ascertain that the user is sending their user Id and not someone's else. Someone may try to fraud using someone else userId. what do you think ?

File: controllers/auth.js (hunk 1 above)
Comment (isVivek99) [reply controllers/auth.js:6]: yes makes sense , i will create an issue ticket if required

**Comment by isVivek99** [comment]
> How will we ascertain that the user is sending their user Id and not someone's else. Someone may try to fraud using someone else userId. what do you think ?



> #1176

I think this was discussed while creating the API contract as well. What we will be using is the deviceId for now which is unique to each users device. What that ID will be is to be decided. We have talked about it being the IMEI number for now

File: controllers/auth.js (hunk 3 above)
Comment (isVivek99) [reply controllers/auth.js]: makes sense

File: controllers/users.js (hunk 4 above)
Comment (isVivek99) [reply controllers/users.js]: ok, i added the space back

File: middlewares/validators/userDeviceInfo.js (hunk 5 above)
Comment (isVivek99) [reply middlewares/validators/userDeviceInfo.js:7]: i think its a `mixed bag`. if we check the `users.js` and `recruiter.js` files inside validators they also have `snake casing`. Also i am following the API contract which was created. [here](https://github.com/Real-Dev-Squad/website-api-contracts/pull/134/files)

File: models/qrCodeAuth.js (hunk 6 above)
Comment (isVivek99) [reply models/qrCodeAuth.js:16]: in this collection `user_id` is the same as `id` in the users collection

File: models/qrCodeAuth.js (hunk 7 above)
Comment (isVivek99) [reply models/qrCodeAuth.js:30]: nope just one

Comment (isVivek99) [reply models/qrCodeAuth.js]: no single user, changed the query

File: controllers/auth.js [hunk 8]
```diff
@@ -1,7 +1,11 @@
 const passport = require("passport");
 const users = require("../models/users");
+const QrCodeAuthModel = require("../models/qrCodeAuth");
 const authService = require("../services/authService");
-
+const { SOMETHING_WENT_WRONG } = require("../constants/errorMessages");
+const DATA_ADDED_SUCCESSFULLY = "User Device Info added successfully!";
+const USER_DATA_ALREADY_PRESENT = "The authentication document has already been created";
+const BAD_REQUEST = "BAD_REQUEST";
```
[the remark this answers is not in the document]
Comment (isVivek99) [reply controllers/auth.js:8]: yes

File: models/qrCodeAuth.js (hunk 6 above)
Comment (isVivek99) [reply models/qrCodeAuth.js:16]: yes its the same id

File: test/integration/qrCodeAuth.test.js
Comment (isVivek99) [reply test/integration/qrCodeAuth.test.js:1]: by mistake, i've removed this file now

File: controllers/auth.js (hunk 1 above)
Comment (isVivek99) [reply controllers/auth.js]: i moved them to constants now

**Comment by heyrandhir** [comment]
Tests are failing. Could you please check @isVivek99

**Review by shreya-mishra — APPROVED** [review]
LGTM

**Comment by isVivek99** [comment]
> How will we ascertain that the user is sending their user Id and not someone's else. Someone may try to fraud using someone else userId. what do you think ?

i have discussed this issue with @shreya-mishra, and we can to a conclusion that we will be adding some sort of authentication in the future either mobile/email verification, for now we will move ahead with this

**Comment by heyrandhir** [comment]
> > How will we ascertain that the user is sending their user Id and not someone's else. Someone may try to fraud using someone else userId. what do you think ?
> 
> i have discussed this issue with @shreya-mishra, and we can to a conclusion that we will be adding some sort of authentication in the future either mobile/email verification, for now we will move ahead with this. Hope this doesn't cause any issue

Lets discuss this with Ankush as well once and take his opinion before proceeding forward. We need to be cautious because once this PR is live in production, there is a possibility that the API might be misused.

**Comment by shreya-mishra** [comment]
> How will we ascertain that the user is sending their user Id and not someone's else. Someone may try to fraud using someone else userId. what do you think ?

This wont be a concern as everything depends on user approving and denying request. The only thing is creation of different documents as many as they want for that also we can set the rate limit on the api. But this wont be a security concern.

**Review by sakshambhatt — DISMISSED** [review]
It would be prudent to apply some common strategies like rate limiting and allowing incoming request from validated instance of app installation (this can be done later).
LGTM :rocket:

**Review by shreya-mishra — APPROVED** [review]

File: models/qrCodeAuth.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+const getUserAuthStatus = async (userId) => {
+  try {
+    const userAuthStatus = await QrCodeAuthModel.where("user_id", "==", userId).get();
+    return userAuthStatus;
+  } catch (err) {
+    logger.log("Could not get", err);
+    throw err;
+  }
+};
```
Comment (heyrandhir) [inline models/qrCodeAuth.js:36]: Aren't we using this anywhere. cant find it used in the controller file .

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: models/qrCodeAuth.js (hunk 9 above)
Comment (isVivek99) [reply models/qrCodeAuth.js]: removed it

**Review by heyrandhir — APPROVED** [review]

**Comment by heyrandhir** [comment]
Please attach the coverage stats.

File: test/integration/auth.test.js [hunk 10]
```diff
@@ -15,6 +14,7 @@ chai.use(chaiHttp);
 const githubUserInfo = require("../fixtures/auth/githubUserInfo")();
 
 describe("auth", function () {
+  beforeEach(async function () {});
```
Comment (heyrandhir) [inline test/integration/auth.test.js:17]: why this beforeEach

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: test/unit/models/qrCodeAuth.test.js [hunk 11]
```diff
@@ -3,16 +3,51 @@ const { expect } = chai;
 
 const cleanDb = require("../../utils/cleanDb");
 const firestore = require("../../../utils/firestore");
-const qrCodeAuthModel = require("../../../models/qrCodeAuth");
-const qrCodeAuth = firestore.collection("QrCodeAuth");
+// const userDeviceInfo = require("../../../models/qrCodeAuth");
+// const userDeviceInfoModel = firestore.collection("QrCodeAuth");
```
Comment (heyrandhir) [inline test/unit/models/qrCodeAuth.test.js:7]: can we remove the commented code if its not needed

File: test/unit/models/qrCodeAuth.test.js [hunk 12]
```diff
@@ -3,16 +3,51 @@ const { expect } = chai;
...
+    it("should store user Id and device info of user for mobile auth", async function () {
+      const userData = userDataArray[0];
+      const { userId } = await users.addOrUpdate(userData);
+
+      const userDeviceInfoData = {
+        user_id: userId,
+        device_info: "TEST_DEVICE_INFO",
+        device_id: "TEST_DEVICE_ID",
+        authorization_status: "NOT_INIT",
+      };
```
Comment (heyrandhir) [inline test/unit/models/qrCodeAuth.test.js:30]: this can be imported form the fixture

File: test/integration/auth.test.js (hunk 10 above)
Comment (isVivek99) [reply test/integration/auth.test.js]: removed it

File: test/unit/middlewares/qrCodeAuthValidator.test.js [hunk 13]
```diff
@@ -1,37 +1,71 @@
...
-    };
+describe("qrCodeAuth", function () {
+  describe("test post call validator", function () {
+    it("Allows request to pass on valid params", async function () {
+      const req = {
+        body: {
+          user_id: "TEST_USER_ID",
+          device_info: "TEST_DEVICE_INFO",
+          device_id: "TEST_DEVICE_ID",
+        },
```
Comment (heyrandhir) [inline test/unit/middlewares/qrCodeAuthValidator.test.js:13]: can we move this to fixture

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: test/unit/models/qrCodeAuth.test.js (hunk 11 above)
Comment (isVivek99) [reply test/unit/models/qrCodeAuth.test.js]: removed it

File: test/unit/models/qrCodeAuth.test.js (hunk 12 above)
Comment (isVivek99) [reply test/unit/models/qrCodeAuth.test.js]: please check now

File: test/unit/middlewares/qrCodeAuthValidator.test.js (hunk 13 above)
Comment (isVivek99) [reply test/unit/middlewares/qrCodeAuthValidator.test.js]: please check now

**Review by heyrandhir — APPROVED** [review]
LGTM with the proposal, that we incorporate two measures. Firstly, we should implement a validator for device information in the request body to ensure that only valid device information is accepted and prevent any random data from being passed through. Secondly, we should also implement rate limiting to prevent a malicious user from creating too many documents in our collection.

**Review by fakhruddinkw — COMMENTED** [review]
Do we need this here?

**Review by fakhruddinkw — APPROVED** [review]

**Review by heyrandhir — APPROVED** [review]

**Review by fakhruddinkw — APPROVED** [review]

**Review by shreya-mishra — APPROVED** [review]
